### ISSUE TITLE
Errors after switching to grudge@main

### ISSUE THREAD

**Issue by MTCam** [illinois-ceesd/mirgecom#617]
MIRGE-Com examples and production drivers appear broken after https://github.com/illinois-ceesd/mirgecom/pull/562/commits/4b78f80d77a8d1dbcac4d7078bbdab61cd8da3d2.

See: https://github.com/illinois-ceesd/mirgecom/runs/5391639189?check_suite_focus=true#step:4:118

Edit: The error message is:

```
File "/home/runner/work/mirgecom/mirgecom/src/grudge/grudge/trace_pair.py", line 500, in cross_rank_trace_pairs
    raise ValueError("Encountered unknown symbolic tag "
ValueError: Encountered unknown symbolic tag '<class 'mirgecom.euler._EulerCVTag'>'. To make this symbolic tag work, use 'grudge.array_context.MPIPyOpenCLArrayContext' and assign this tag a numerical value via its comm_tag_to_mpi_tag attribute.
```

**Comment by MTCam** [comment]
After a little digging, it appears the fix is to do something like this:
```
114    actx_class = get_reasonable_array_context_class(lazy=lazy, distributed=True)                                                                  
115    if lazy:                                                                                                                                      
116        actx = actx_class(comm, queue, mpi_base_tag=12000)                                                                                        
117    else:                                                                                                                                         
118        actx = actx_class(comm, queue,                                                                                                            
119                allocator=cl_tools.MemoryPool(cl_tools.ImmediateAllocator(queue)),                                                                
120                force_device_scalars=True,                                                                                                        
121                comm_tag_to_mpi_tag={                                                                                                             
122                    _WaveStateTag: 1234,                                                                                                          
123                    })                                                                                                                            
```
This is sort of a problem for fluid operators as we do not have access to the communication tags at the driver level (i.e. the level at which the array context is created), and pulling them up to the driver level is cumbersome and unnatural.

Communication required by the fluid operators (for example) is performed at the fluid operator level, and the tags for those communications are defined by those operator modules.  I think maybe we need a module from which to pull comm tags, and we can just pack them all into a dictionary and pass the entire bag of them to the array context class constructor.

@inducer @matthiasdiener

**Comment by matthiasdiener** [comment]
This is caused by the changes in https://github.com/inducer/grudge/pull/183. Instead of aborting when not encountering a symbolic tag -> numeric tag mapping, could we just choose a default numeric value @inducer?

**Comment by matthiasdiener** [comment]
> Instead of aborting when not encountering a symbolic tag -> numeric tag mapping in the eager case, could we just choose a default numeric value @inducer?

https://github.com/inducer/grudge/pull/222 implements this idea.

**Comment by inducer** [comment]
Sorry about the breakage, I should have warned you guys. When finishing up https://github.com/inducer/grudge/pull/183, I came across the correctness problems described below, and the only way I could think of that would address this involved an interface change; I decided on the one you see in `grudge@main`. I've since gone back and forth on whether it's possible to offer correct semantics in the eager case without burdening the user with providing the tag mapping. I think I've convinced myself that it *is* possible, so my current guess is that `comm_tag_to_mpi_tag` should only be required temporarily.

> Instead of aborting when not encountering a symbolic tag -> numeric tag mapping in the eager case, could we just choose a default numeric value @inducer?

We can't. It's not correct semantics for grudge to do that. You wouldn't expect communication sent with one tag to match a receive with another, right? https://github.com/inducer/grudge/pull/222 also does not qualify since `a != b` does not imply `hash(a) != hash(b)`.

That said, it's easy to request this same behavior from the user end by passing a `defaultdict` for `comm_tag_to_mpi_tag`, this avoids having to enumerate all tags in use as long as all send/receive pairs are sufficiently distinguished by program order.

@matthiasdiener My thought is that, for each (directed) point-to-point communication pair, each communication partner knows that whether a tag mapping is available. If it isn't, we'll send the symbolic tag and its mapping (unilaterally decided by the sender) on a special tag and start communicating as normal thereafter. Could you look into implementing that?

tl;dr: Pass a `defaultdict` for now, I think we can do without `comm_tag_to_mpi_tag` ultimately (and it should be quick to implement that path).

**Comment by inducer** [comment]
As described in https://github.com/inducer/grudge/pull/222/files#r817836129, I'd be OK with merging that as a temporary stopgap.

**Comment by MTCam** [comment]
> As described in [inducer/grudge#222 (files)](https://github.com/inducer/grudge/pull/222/files#r817836129), I'd be OK with merging that as a temporary stopgap.

if inducer/grudge#222 is not an acceptable/permanent repair for this, then I would much prefer the *MIRGE-Com* -side fix that is done in https://github.com/illinois-ceesd/mirgecom/pull/562/commits/94ac2b0c3e5b853c22d7983b441d62a7d11f67ac.

**Comment by inducer** [comment]
My point is that (I think) we can get by without needing mirgecom-side changes with some (minor) additional work on the grudge end. As a result, I think it's OK to be temporarily incorrect and avoid inflicting there-and-back changes on mirgecom.

**Comment by MTCam** [comment]
> My point is that (I think) we can get by without needing mirgecom-side changes with some (minor) additional work on the grudge end. As a result, I think it's OK to be temporarily incorrect and avoid inflicting there-and-back changes on mirgecom.

Understood and agreed.   You are saying fix it up temporarily with inducer/grudge#222 and then for real after some thought.  That's fine with me.   I can confirm that  inducer/grudge#222 fixes stuff on the MIRGE-Com side.

**Comment by MTCam** [comment]
> ... avoid inflicting there-and-back changes on mirgecom.

FWIW, I think https://github.com/illinois-ceesd/mirgecom/pull/562/commits/94ac2b0c3e5b853c22d7983b441d62a7d11f67ac is a positive change that would not need backing out.

**Comment by inducer** [comment]
I think we can get by without a user-supplied `comm_tag_to_mpi_tag`, so you can just delete that and the bits that create those mappings.

**Comment by inducer** [comment]
See the changes I just pushed to https://github.com/inducer/grudge/pull/222.

**Comment by MTCam** [comment]
inducer/grudge#222 appears to fix or work around this issue now 👍
